pkg/scalers/splunk_scaler.go

```diff
@@ -0,0 +1,130 @@
+package scalers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+
+	"github.com/go-logr/logr"
+	v2 "k8s.io/api/autoscaling/v2"
+	"k8s.io/metrics/pkg/apis/external_metrics"
+
+	"github.com/kedacore/keda/v2/pkg/scalers/scalersconfig"
+	"github.com/kedacore/keda/v2/pkg/scalers/splunk"
+	kedautil "github.com/kedacore/keda/v2/pkg/util"
+)
+
+// SplunkScaler assigns struct data pointer to metadata variable
+type SplunkScaler struct {
+	client     *splunk.Client
+	metricType v2.MetricTargetType
+	metadata   SplunkMetadata
+	logger     logr.Logger
+}
+
+// SplunkMetadata Metadata used by KEDA to search Splunk events and scale
+type SplunkMetadata struct {
+	APIToken        string `keda:"name=apiToken,        order=authParams, optional"`
+	Password        string `keda:"name=password,        order=authParams, optional"`
+	Username        string `keda:"name=username,        order=triggerMetadata"`
+	Host            string `keda:"name=host, 	        order=triggerMetadata"`
+	UnsafeSsl       bool   `keda:"name=unsafeSsl,       order=triggerMetadata, optional"`
+	TargetValue     int    `keda:"name=targetValue, 	order=triggerMetadata"`
+	ActivationValue int    `keda:"name=activationValue, order=triggerMetadata"`
+	SavedSearchName string `keda:"name=savedSearchName, order=triggerMetadata"`
+	ValueField      string `keda:"name=valueField,      order=triggerMetadata"`
+	triggerIndex    int
+}
+
+func NewSplunkScaler(config *scalersconfig.ScalerConfig) (Scaler, error) {
+	metricType, err := GetMetricTargetType(config)
+	if err != nil {
+		return nil, fmt.Errorf("error getting scaler metric type: %w", err)
+	}
+
+	meta, err := parseSplunkMetadata(config)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing Splunk metadata: %w", err)
+	}
+
+	client, err := splunk.NewClient(&splunk.Config{
+		APIToken:  meta.APIToken,
+		Password:  meta.Password,
+		Username:  meta.Username,
+		Host:      meta.Host,
+		UnsafeSsl: meta.UnsafeSsl,
+	}, config)
+	if err != nil {
+		return nil, err
+	}
+
+	return &SplunkScaler{
+		client:     client,
+		metricType: metricType,
+		logger:     InitializeLogger(config, "splunk_scaler"),
+		metadata:   *meta,
+	}, nil
+}
+
+func (s *SplunkScaler) Close(context.Context) error {
+	return nil
+}
+
+func parseSplunkMetadata(config *scalersconfig.ScalerConfig) (*SplunkMetadata, error) {
+	meta := &SplunkMetadata{}
+	meta.triggerIndex = config.TriggerIndex
+	if err := config.TypedConfig(meta); err != nil {
+		return nil, fmt.Errorf("error parsing splunk metadata: %w", err)
+	}
+
+	_, err := url.ParseRequestURI(meta.Host)
+	if err != nil {
+		return meta, errors.New("invalid value for host. Must be a valid URL such as https://localhost:8089")
+	}
+
+	if meta.Username == "" {
+		return meta, errors.New("no username given")
+	}
+
+	if meta.SavedSearchName == "" {
+		return meta, errors.New("no savedSearchName given")
+	}
+
+	if meta.ValueField == "" {
+		return meta, errors.New("no valueField given")
+	}
```

Comment:
nit: given these three `Username`, `SavedSearchName`, `ValueField` are not marked as `optional`, it should be safe to remove the check as it's performed by `TypedConfig()`

Reply:
That just validates the keys though, not the values?

Reply:
I have double-mislead you during our discussion, my apologies.

After checking the code, compulsory parameters (those not marked as `optional`), are indeed checked for empty string too. The algorithm returns true only in case the param is present in one of the desired maps and its value is not `""`
https://github.com/kedacore/keda/blob/ec525738d8c42a3c622be1fd78977b0753bd3e69/pkg/scalers/scalersconfig/typed_config.go#L412-L414

So after claiming it should be safe to remove these checks, then instructed you to keep the checks, I now finally *know* it is safe to remove the checks :)

Reply:
Ah nice, thanks for the catch. I've removed these redundant checks